decidim-core/db/migrate/20170605162500_add_hierarchy_to_scopes.rb

```diff
@@ -0,0 +1,60 @@
+# frozen_string_literal: true
+
+class AddHierarchyToScopes < ActiveRecord::Migration[5.0]
+  def self.up
+    # schema migration
+    create_table :decidim_scope_types do |t|
+      t.references :decidim_organization, foreign_key: true, index: true
+      t.jsonb :name, null: false
+      t.jsonb :plural, null: false
+    end
+
+    change_table :decidim_scopes do |t|
+      t.remove_index :name
+      t.jsonb :name2, null: false
+      t.references :scope_type, foreign_key: { to_table: :decidim_scope_types }, index: true
+      t.references :parent, foreign_key: { to_table: :decidim_scopes }
+      t.string :code
+      t.jsonb :metadata, default: {}, null: false
+      t.boolean :deprecated, default: false, null: false
+      t.integer :part_of, array: true, default: [], null: false
+      t.index :part_of, using: "gin"
+    end
+
+    # post migration data fixes
+    Decidim::Scope.all.find_each do |s|
```

Comment:
Maybe you can store all the data in an in-memory variable instead - doing it this way looks a bit hacky, doesn't it?

Wrap the whole thing inside a transaction and it's good to go!